relay-server/src/utils/garbage.rs

```diff
@@ -0,0 +1,82 @@
+use std::sync::mpsc;
+
+pub struct GarbageDisposal<T> {
+    tx: mpsc::Sender<T>,
+}
+
+impl<T: Send + 'static> GarbageDisposal<T> {
+    pub fn new() -> Self {
+        let (tx, rx) = mpsc::channel();
+
+        std::thread::spawn(move || {
+            relay_log::info!("Start garbage collection thread");
+            while let Ok(object) = rx.recv() {
```

Comment:
Consider adding a cancellation token (e.g. CondVar) that stops the loop (shutting down the thread) when `GarbageDisposal` drops.

Reply:
This thread will already stop once the sender (`GarbageDisposal`) is dropped, that's what `while let Ok(_)` does since `rx.recv()` will return `Err` if the sender is dropped.  So like, `rx` is this guard you're asking for I think?

Reply:
Dropping `GarbageDisposal` does stop the loop (by dropping `tx`), but it does not wait for the spawned thread to actually finish. I've fixed that in a more recent PR, but it does make the code less readable.

Reply:
when do you need to wait for the thread to finish?  If this is for tests only I'd be tempted to split the new function up in a `new` for prod and `new_with_handle` for the tests, the test can then do the waiting on the handle but prod doesn't need to have all this extra code complexity.